lib/axon/loop.ex

```diff
@@ -384,15 +384,17 @@ defmodule Axon.Loop do
 
     init_fn = fn ->
       %{
+        model_state: model_state,
```

Comment:
Is `init_fn` jitted? If so, we have the same issue that we are passing a `EXLA.Backend` tensor to `defn` behind the scenes, which means that it will be converted to binary and inlined into the code. Is this what we want? It feels like we should be raising on said implicit conversions. Should we change Nx to raise?

In any case, I like the direction of moving the concern to `init_fn`. What if we actually pass the model state as argument to `init_fn`? This would build on top of this change and address all of my concerns.

If we do so, I would also change Nx to raise if we try to convert any non binary tensor to an expression (because it almost always means that we are passing something as a closure when we should not).

WDYT?

Reply:
> If we do so, I would also change Nx to raise if we try to convert any non binary tensor to an expression (because it almost always means that we are passing something as a closure when we should not).

I added this check and caught bugs on our EXLA.Backend implementation, so I would say this is the way to go!